OpenSim/Actuators/PolynomialPathFitter.cpp

```diff
@@ -1310,6 +1274,209 @@ void PolynomialPathFitter::removeMomentArmColumns(TimeSeriesTable& momentArms,
+                momentArmMet = false;
+                break;
+            }
+        }
+
+        if ((pathLengthMet && momentArmMet) ||
+                order == get_maximum_polynomial_order()) {
+            break;
+        }
+        ++order;
+    }
+}
+
+void PolynomialPathFitter::fitCoefficientsStepwiseRegression(
+        const SimTK::Matrix& coordinates, const SimTK::Vector& b, int order,
+        SimTK::Vector& coefficients) const {
+
+    // Preliminaries.
+    int numTimes = coordinates.nrow();
+    int numCoordinates = coordinates.ncol();
+    int numCoefficients = choose(numCoordinates + order, order);
+
+    // Construct the full 'A' matrix.
+    SimTK::Vector dummyCoefficients(numCoefficients, 1.0);
+    MultivariatePolynomialFunction dummyFunction(dummyCoefficients,
+            numCoordinates, order);
+    SimTK::Matrix Afull(numTimes * (numCoordinates + 1),
+            numCoefficients, 0.0);
+    for (int itime = 0; itime < numTimes; ++itime) {
+        Afull(itime, 0, 1, numCoefficients) =
+                dummyFunction.getTermValues(
+                    coordinates.row(itime).getAsVector());
+
+        for (int ic = 0; ic < numCoordinates; ++ic) {
+            SimTK::Vector termDerivatives =
+                dummyFunction.getTermDerivatives({ic},
+                    coordinates.row(itime).getAsVector())
+                    .negate();
+            Afull((ic+1)*numTimes + itime, 0, 1, numCoefficients) =
+                    termDerivatives;
+        }
+    }
+
+    // Manage the coefficient indexes that will be included in the final 
+    // polynomial. The "out" indexes are the indexes that are not included in
+    // the final polynomial, which is initialized to all indexes. The "keep" 
+    // indexes are the indexes that are included in the final polynomial, which
+    // is initialized to an empty vector.
+    std::vector<int> outIndexes;
+    outIndexes.reserve(numCoefficients);
+    for (int i = 0; i < numCoefficients; ++i) {
+        outIndexes.push_back(i);
+    }
+    std::vector<int> keepIndexes;
+    SimTK::Vector x_sol;
+    while (true) {
+
+        // Initialize the 'A' matrix. 
+        // The number of terms in the polynomial is the number of "kept" 
+        // coefficients plus one. 
+        int numTerms = static_cast<int>(keepIndexes.size()) + 1;
+        SimTK::Matrix A(numTimes * (numCoordinates + 1), numTerms, 0.0);
+        SimTK::Vector x(numTerms, 0.0);
+
+        // First, set the columns of 'A' for the "kept" coefficients.
+        int icol = 0;
+        for (const auto& ki : keepIndexes) {
+            A.updCol(icol++) = Afull.col(ki);
+        }
+
+        // Next, loop through all of the remaining "out" coefficients and fit 
+        // the polynomial. We will keep the coefficient that results in the
+        // smallest RMS error.
+        SimTK::Real bestError = SimTK::Infinity;
+        int bestIndex = -1;
+        for (const auto& oi : outIndexes) {
```

Comment:
Will this not become extremely expensive as `A` starts getting some nonzero size? Or is computational time not so important during this phase?

Reply:
This ends up not being as expensive as it seems. The biggest bottleneck in the fitter is (by far) calculating path length and moment arm values, especially if you use a lot of sample points. But the time can be reduced significantly by sampling the model coordinates in an efficient way.